#107830854 Upgrade user account

oeyiowuawi opened the pull request:
#### What does this PR do?

Enables user to upgrade to a different plan
#### Description of Task to be completed?
- add user plan model
- user plan defaults to novice on signup
- display current plan on view
- remove `plan` and `expiry_date` from user model
- remove user plan management methods
- remove upgrade plan spec
#### How should this be manually tested?
- user logs in
- clicks on Account Upgrade on the sidebar
- user sees different plans and makes a choice
- user clicks `pay with card` and enters card details
- user is provided with a notification status
#### What are the relevant pivotal tracker stories?
- Finishes  #107830854 
#### Screenshots

![](http://cdn.kagumba.com/music_library/plan.png)

![](http://cdn.kagumba.com/music_library/plan2.png)

![](https://www.dropbox.com/s/9380ccw8dlyxssg/Screen%20Shot%202016-04-13%20at%2011.12.54%20AM.png?raw=1)

![](https://www.dropbox.com/s/ztj94w1qzmjpoge/Screen%20Shot%202016-04-13%20at%2011.16.48%20AM.png?raw=1)

Commit:
test(upgrade plans): initial test scenario

Commit:
feat(upgrade): add user pricing plan view

Commit:
chore(upgrade): add taskee pricing cards

Commit:
feat(tasker upgrade): add checks for tasker plan before creating a task

Commit:
Merge branch 'ft-upgrade-account-107830854' of https://github.com/andela/workdey into ft-upgrade-account-107830854

Commit:
feat(tasker upgrade): add necessary migrations to the db

Commit:
Merge branch 'master' into ft-upgrade-account-107830854

Commit:
feat(novice): add check for novice users

Commit:
feat(upgrade plan): integrate stripe checkout

charge base on user plan type
check subscription before creating a task

Commit:
feat(upgrade plan): add user_plan model

add `valid_until`, `user_id` and `tasks_counter` to
track user plan

Commit:
feat(upgrade plan): upgrade account with stripe

 - add user plan model
 - user plan defaults to novice on signup
 - display current plan on view
 - remove `plan` and `expiry_date` from user model
 - remove user plan management methods
 - remove upgrade plan spec

 Closes #107830854

Commit:
fix(upgrade account): resolve merge conflict

 - merge changes from upstream
 - add seed data for user plan

Commit:
chore(maintain): fix style violations

Commit:
chore(user migrations): add missing migration files

Commit:
chore(server): add puma to production

Commit:
chore(stripe): add test key for production

Commit:
chore(upgrade plan): add stripe environment variables

Commit:
chore(upgrade plans): check current plan of user

 - change card title to capitalized
 - show current user plan and prevoius plans

Commit:
fix(user plnan): use  in

codesword commented on app/controllers/user_plans_controller.rb:
```diff
@@ -0,0 +1,34 @@
+
+  private
+
+  def check_amount
```
I wonder why this method is called `check_amount` if it is setting both the amount and description of a plan.

codesword commented on app/controllers/user_plans_controller.rb:
```diff
@@ -0,0 +1,34 @@
+    customer = Stripe::Customer.create(email: params[:stripeEmail],
+                                       source: params[:stripeToken])
+
+    Stripe::Charge.create(
```
You could extract this code block into a method and avoid using instance variable by doing something like this.

``` ruby
def charge_customer_through_stripe(customer_id)
    amount, description = amount_and_description_by_plan
    Stripe::Charge.create(
       customer: customer_id,
       amount: amount,
       description: description,
      currency: "usd"
     )
end

def amount_and_description_by_plan
   if params[:plan] == "medial"
       [2000, "Workdey medial user"]
   elsif params[:plan] == "maestro"
        [3000, "Workdey maestro user"]
   end
end
```

codesword commented on app/models/user.rb:
```diff
@@ -81,6 +82,18 @@ def self.get_taskees_by_task_name(keyword, user_email = nil)
     taskees.where("email != ?", user_email)
   end
 
+  def novice?
```
👍 On creating all these utility methods. Now it's easy to know if a user is in a specific plan or not.

codesword commented on app/models/user_plan.rb:
```diff
@@ -0,0 +1,12 @@
+  belongs_to :user, foreign_key: :user_id
+
+  def self.subscribe_user(user_id, plan)
+    find_by(user_id: user_id).update_attributes(active_until: (Date.today + 1.year),
```
Please fix the houndCI issue.

codesword commented on app/views/user_plans/index.html.erb:
```diff
@@ -0,0 +1,177 @@
+      <section class="col s12">
+        <div class="row">
+          <h2 class="header">User Plans</h2>
+          <% if current_user.user_type == "taskee" %>
```
These plan cards could be a partial which could be shared by all the plans. Which means both the plan title, price, plan benefits(receive 10 tasks, earn reputation etc) should come from the database.

Commit:
Merge branch 'master' into ft-upgrade-account-107830854

Commit:
chore(user plan model): fix hound ci violations

 - fix Gemfile.lock file
 - fix style violations on user plan model

[Finishes #107830854]

codesword commented:
This looks good. I will be merging now.
